src/ganted2radius.go

```diff
@@ -0,0 +1,198 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/klauspost/compress/zstd"
+	"golang.org/x/net/context"
+	"layeh.com/radius"
+	"layeh.com/radius/rfc2865"
+	"layeh.com/radius/rfc2866"
+)
+
+func compressFile(filename string) error {
+	// open the file
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	// create the compressed file, err if it already exists
+	compressedFilename := filename + ".zst"
+	compressedFile, err := os.Create(compressedFilename)
+	if err != nil {
+		return err
+	}
+	defer compressedFile.Close()
+	// create the zstd writer
+	zw, err := zstd.NewWriter(compressedFile)
+	if err != nil {
+		return err
+	}
+	// copy the file to the zstd writer
+	_, err = io.Copy(zw, file)
+	if err != nil {
+		return err
+	}
+	// flush the zstd writer
+	if err := zw.Close(); err != nil {
+		return err
+	}
+	// remove the original file
+	if err := os.Remove(filename); err != nil {
+		return err
+	}
+	return nil
+}
+
+func archiveLogs(logDir string) error {
+	logPattern := regexp.MustCompile(`^access-\d{14}\.log$`)
+	// compress all access-<datetime>.log files to access-<datetime>.log.zst
+	err := filepath.Walk(logDir, func(path string, info os.FileInfo, err error) error {
```

Comment:
I'm a little concerned about the low performance of this operation when the number of logs is relatively large.
The solutions I can think of:
1) List all files and filter them first. But it seems that Go doesn't have functions to do this directly, or I miss something.
2) Use goroutine to speed up this.
3) Assume there is only one `access-<datetime>.log` that needs to be compressed.

Reply:
Emmm... Is `filepath.Walk()` slow? Maybe we could use `filepath.WalkDir()` and check filename first to avoid unnecessary `lstat()`.

Also I think that the compress part could be put into a goroutine (if it is slow), as we don't need to wait for compression to be done before accounting. Another approach is to do rotate only in the go program, and if there're already 24 files (hours) inside we call an external script to do extra logrotating and compressing like https://github.com/ustclug/light-server/blob/e1aa99d4790d4dd163fbe8771aa3093d40078548/.docker/rootfs/etc/periodic/hourly/squid2radius#L7-L9.

Reply:
Resolved at new commit. Just rotate when accounting, and archive(concatenate and compress) every 24 log files (daily).